File: backend/api/api_perfiles.py
from fastapi import FastAPI, HTTPException
import asyncpg
from contextlib import asynccontextmanager
import os
from pydantic import BaseModel
from typing import List

# =================================================================
# CONFIGURACIÓN DE BASE DE DATOS
# =================================================================
DB_CONFIG = {
    'user': 'postgres',
    'password': 'tu_password',
    'database': 'ecosystems_db',
    'host': os.getenv('DB_HOST', '127.0.0.1'),
    'port': 5432
}

pool = None

@asynccontextmanager
async def lifespan(app: FastAPI):
    global pool
    print("Iniciando API de Perfiles de Cultivo...")
    pool = await asyncpg.create_pool(**DB_CONFIG)
    yield
    if pool:
        await pool.close()

app = FastAPI(
    title="API Perfiles EcoSystems", 
    description="Microservicio CRUD para gestionar umbrales de humedad por cultivo (US-04)",
    lifespan=lifespan
)

# =================================================================
# MODELOS PYDANTIC
# =================================================================

class PerfilCultivoBase(BaseModel):
    nombre_cultivo: str
    humedad_min_prc: int
    humedad_max_prc: int

class PerfilCultivo(PerfilCultivoBase):
    id_perfil: int

# =================================================================
# ENDPOINTS REST (CRUD)
# =================================================================

@app.get("/api/perfiles", response_model=List[PerfilCultivo])
async def obtener_perfiles():
    """ Lista todos los perfiles de cultivo configurados. """
    if not pool:
        raise HTTPException(status_code=503, detail="BD no disponible")
    async with pool.acquire() as conn:
        registros = await conn.fetch("SELECT * FROM perfil_cultivo ORDER BY id_perfil ASC")
        return [dict(r) for r in registros]

@app.post("/api/perfiles", response_model=PerfilCultivo)
async def crear_perfil(perfil: PerfilCultivoBase):
    """ Crea un nuevo perfil de cultivo. """
    if not pool:
        raise HTTPException(status_code=503, detail="BD no disponible")
    
    query = """
        INSERT INTO perfil_cultivo (nombre_cultivo, humedad_min_prc, humedad_max_prc)
        VALUES ($1, $2, $3) RETURNING id_perfil
    """
    try:
        async with pool.acquire() as conn:
            id_nuevo = await conn.fetchval(query, perfil.nombre_cultivo, perfil.humedad_min_prc, perfil.humedad_max_prc)
        return {**perfil.model_dump(), "id_perfil": id_nuevo}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@app.delete("/api/perfiles/{id_perfil}")
async def eliminar_perfil(id_perfil: int):
    """ Elimina un perfil de cultivo existente. """
    if not pool:
        raise HTTPException(status_code=503, detail="BD no disponible")
        
    query = "DELETE FROM perfil_cultivo WHERE id_perfil = $1 RETURNING id_perfil"
    try:
        async with pool.acquire() as conn:
            borrado = await conn.fetchval(query, id_perfil)
            if not borrado:
                raise HTTPException(status_code=404, detail="Perfil no encontrado")
        return {"status": "success", "message": f"Perfil {id_perfil} eliminado"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: backend/node-api/api/api_perfiles.js
const express = require('express');
const { Pool } = require('pg');
const cors = require('cors');

const app = express();
app.use(express.json());
app.use(cors());

const PORT = process.env.PORT || 8003;
const DB_CONFIG = {
    user: 'postgres', password: 'tu_password', database: 'ecosystems_db', host: process.env.DB_HOST || '127.0.0.1', port: 5432
};
const pool = new Pool(DB_CONFIG);

app.get('/api/perfiles', async (req, res) => {
    try {
        const { rows } = await pool.query("SELECT * FROM perfil_cultivo ORDER BY id_perfil ASC");
        res.json(rows);
    } catch (e) {
        res.status(500).json({ detail: e.message });
    }
});

app.post('/api/perfiles', async (req, res) => {
    const { nombre_cultivo, humedad_min_prc, humedad_max_prc } = req.body;
    try {
        const query = `
            INSERT INTO perfil_cultivo (nombre_cultivo, humedad_min_prc, humedad_max_prc)
            VALUES ($1, $2, $3) RETURNING id_perfil
        `;
        const { rows } = await pool.query(query, [nombre_cultivo, humedad_min_prc, humedad_max_prc]);
        res.json({ id_perfil: rows[0].id_perfil, nombre_cultivo, humedad_min_prc, humedad_max_prc });
    } catch (e) {
        res.status(500).json({ detail: e.message });
    }
});

app.delete('/api/perfiles/:id_perfil', async (req, res) => {
    const id_perfil = parseInt(req.params.id_perfil);
    try {
        const query = "DELETE FROM perfil_cultivo WHERE id_perfil = $1";
        const { rowCount } = await pool.query(query, [id_perfil]);
        
        if (rowCount === 0) return res.status(404).json({ detail: "Perfil no encontrado" });
        
        res.json({ status: "success", message: `Perfil ${id_perfil} eliminado` });
    } catch (e) {
        res.status(500).json({ detail: e.message });
    }
});

app.listen(PORT, () => console.log(`API Perfiles corriendo en puerto ${PORT}`));